lib/membrane/video_compositor/handler.ex

```diff
@@ -0,0 +1,77 @@
+defmodule Membrane.VideoCompositor.Handler do
+  @moduledoc """
+  Module defining behaviour of handlers.
+
+  Implementing handler allows to provide custom implementation and
+  react to various events, among others by setting a new scene and
+  the inner custom state.
+  """
+  alias Membrane.Pad
+  alias Membrane.VideoCompositor
+  alias Membrane.VideoCompositor.Handler.InputProperties
+  alias Membrane.VideoCompositor.Scene
+
+  @typedoc """
+  Type of user-managed inner state of the handler.
+  """
+  @type state :: any()
+
+  @typedoc """
+  Contains state of VC before handling event invoking callback.
+  """
+  @type context :: %{
+          scene: Scene.t(),
+          inputs: inputs(),
+          next_frame_pts: Membrane.Time.non_neg_t(),
+          scenes_queue: [{start_pts :: Membrane.Time.non_neg_t(), new_scene :: Scene.t()}]
+        }
+
+  @typedoc """
+  Describe all VC input videos used in composition.
+  """
+  @type inputs() :: %{Pad.ref_t() => InputProperties.t()}
+
+  @typedoc """
+  Type of a valid return value from the callback. By returning this type,
+  the scene will be changed immediate, i.e. at the moment when the event happens.
+  """
+  @type immediate_callback_return :: {scene :: Scene.t(), state :: state()}
+
+  @typedoc """
+  Type of a valid return value from callback allowing to pick start time of a new scene.
+  """
+  @type timed_callback_return ::
+          {{scene :: Scene.t(), start_pts :: Membrane.Time.non_neg_t()}, state :: state()}
+
+  @doc """
+  Callback invoked upon initialization of Video Compositor.
+  """
+  @callback handle_init(init_options :: VideoCompositor.init_options()) :: state()
+
+  @doc """
+  Callback invoked upon change of VC `t:inputs()`.
+
+  `inputs` changing input videos:
+  - video added
+  - video removed
+  - video stream format change
```

Comment:
What about metadata? 
Example: there is an input track that is at first a camera stream and then it changes into a screen share.
Do I have to write custom messages on metadata changes but in this case how to change the context for future calls?

Reply:
The `metadata` is passed as an input pad option.
If sth can dynamically change in your application, passing it as a linking option and allowing to modify it, doesn't sound like a good idea. If we allow changing it here and sth crashes, error logs will show different metadata in membrane contexts and VC logs, making it more difficult to debug.

In the case of screen sharing, I would do it like this:
1. Pass the initial "sharing state" option as `metadata` in `pad options`, like this: `via_in(.., options: [..., metadata: %{initial_sharing_state: :camera}]) |> get_child(:video_compostior)`
2. On `handle_inputs_change` check if you have saved the initial option in your `state`. If not, put it in your `state`. Generate a `scene` based on your `state`
3. In case of changing the "sharing state" without changing the input pad, send a message into the VC process and handle it in `handle_info`, where you update the pad "sharing state" in your `state` and generate a `scene`.

```elixir
defmodule VideoCallWithScreenShareHandler do
  @moduledoc false

  @behaviour Membrane.VideoCompositor.Handler

  @impl true
  def handle_init(init_options) do
    %{sharing_state: %{}}
  end

  @impl true
  def handle_inputs_change(inputs, _ctx, state = %{sharing_state: sharing_state}) do
    state = %{
      state
      | sharing_state:
          Map.new(
            inputs,
            fn {pad, pad_properties} ->
              Map.get(sharing_state, pad, pad_properties.metadata.initial_pad_sharing_state)
            end
          )
    }

    {get_scene(state), state}
  end

  @impl true
  def handle_info(msg, ctx, state) do
    case msg do
      {:sharing_state, pad, pad_sharing_state} ->
        state = %{state | sharing_state: Map.put(state.sharing_state, pad, pad_sharing_state)}
        {get_scene(state), state}
      _else ->
        {ctx.current_scene, state}
    end
  end
end
```